xla/service/gpu/autotuning/gemm_algorithm_picker.cc

```diff
@@ -98,6 +98,14 @@ class GemmAutotuner {
   explicit GemmAutotuner(const AutotuneConfig& autotune_config)
       : autotune_config_(autotune_config) {}
 
+  ~GemmAutotuner() {
+    if (stream_ != nullptr) {
+      if (auto blas = stream_->parent()->AsBlas()) blas->ResetStream();
+    }
```

Comment:
I think what we actually want is that the stream notifies the Blas instance when it gets out of scope.

We already have a similar mechanism for DNN (For example: https://github.com/openxla/xla/blob/c2348031b2f8c7c1719eace85f112171f0e06d0a/xla/stream_executor/cuda/cuda_executor.cc#L499).

I'm not sure what architecture we should strive for here, so I'm going to ask a colleague who knows more about stream_executor and will come back to you ASAP.

Reply:
Another thought: It seems we call `SetStream` in each call to `ROCMBlas::DoBlasInternalImpl`.
Couldn't we just reset the stream back to zero at the end of `ROCMBlas::DoBlasInternalImpl`?

This would be more in line with the API and less spaghetti-like communication between parts of the code base. I'm not sure though whether this has perfomance implications.

Reply:
oh I think I like the first idea: we can add such a 'listener' function NotifyStreamDestroyed to Blas interface. Than we can get rid of a specific destructor for GemmAlgorithmPicker. I will try it out. Setting the stream to 0 after each call to DoBlasInternalImpl sounds less generic to me because we can potentially do more cleanup tasks in **NotifyStreamDestroyed** function (like for example, resetting user workspace bound to this stream?)

Reply:
> Setting the stream to 0 after each call to DoBlasInternalImpl sounds less generic to me because we can potentially do more cleanup tasks in NotifyStreamDestroyed function (like for example, resetting user workspace bound to this stream?)

It's true that the listener approach is more generic, but it's also way more complicated. Therefore if the "Setting the stream to 0 after each call to DoBlasInternalImpl" has no serious downside, then I would highly prefer this approach over the other.

That of course doesn't mean that we can't introduce the listener approach in the future if there is a functionality that needs it - together with some proper testing of this mechanism.

Reply:
Sounds good. Thank you for making the change.